tests/test_measure.py

```diff
@@ -0,0 +1,116 @@
+    """Test that failing to specify a measurement
+    raises an exception"""
+    dev = qml.device("default.qubit", wires=2)
+
+    @qml.qnode(dev)
+    def circuit(x):
+        qml.RX(x, wires=0)
+        return qml.PauliY(0)
+
+    with pytest.raises(QuantumFunctionError, match="does not have the measurement"):
+        res = circuit(0.65)
+
+
+class TestExpval:
+    """Tests for the new expval function"""
+
+    def test_value(self, tol):
+        """Test that the new expval interface works"""
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev)
+        def circuit(x):
+            qml.RX(x, wires=0)
+            return qml.expval(qml.PauliY(0))
+
+        x = 0.54
+        res = circuit(x)
+        expected = -np.sin(x)
+
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+    def test_not_an_observable(self):
+        """Test that a QuantumFunctionError is raised if the provided
+        argument is not an observable"""
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.RX(0.52, wires=0)
+            return qml.expval(qml.CNOT(wires=[0, 1]))
+
+        with pytest.raises(QuantumFunctionError, match="CNOT is not an observable"):
+            res = circuit()
+
+
+class TestDeprecatedExpval:
+    """Tests for the deprecated expval attribute getter.
+    Once fully deprecated, this test can be removed"""
+
+    def test_value(self, tol):
+        """Test that the old expval interface works,
+        but a deprecation warning is raised"""
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev)
+        def circuit(x):
+            qml.RX(x, wires=0)
+            return qml.expval.PauliY(0)
+
+        x = 0.54
+        with pytest.warns(DeprecationWarning, match="is deprecated"):
+            res = circuit(x)
+
+        expected = -np.sin(x)
+
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+    def test_not_an_observable(self):
+        """Test that an attribute error is raised if the provided
+        attribute is not an observable when using the old interface"""
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.RX(0.52, wires=0)
+            return qml.expval.RX(0)
+
+        with pytest.warns(DeprecationWarning, match="is deprecated"):
+            with pytest.raises(AttributeError, match="RX is not an observable"):
+                res = circuit()
+
+    def test_not_an_operation(self):
+        """Test that an attribute error is raised if an
+        observable doesn't exist using the old interface"""
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.RX(0.52, wires=0)
+            return qml.expval.R(0)
+
+        with pytest.warns(DeprecationWarning, match="is deprecated"):
+            with pytest.raises(AttributeError, match="has no attribute 'R'"):
```

Comment:
Wouldn't we want it to give a better warning message? i.e., instead of `attribute`, something like `has no observable 'R'`

Reply:
Good point. I actually checked what error message is currently raised in PennyLane if you call `qml.expval.R`, and copied it word for word.

Reply:
done